### PR TITLE
Add different sizes of metal lumps

### PR DESCRIPTION
**Pull request by juur**
#### Summary
Content "Add different sizes and recipes for various lumps of metal"

#### Purpose of change

Allow crafting as a mechanism to go from scrap to 1L to 5L to 10L and back again. As part of this, normalise some missing 1/5/10L ingots.

#### Describe the solution

Added missing recipes, items and uncrafts. 

#### Testing

Loading of game, visual inspection inside crafting window, and tests via spawning.

### CODE REVIEW

File: data/json/recipes/other/materials.json [hunk 1]
```diff
@@ -239,6 +318,45 @@
...
+    "type": "recipe",
+    "activity_level": "LIGHT_EXERCISE",
+    "category": "CC_OTHER",
+    "subcategory": "CSC_OTHER_MATERIALS",
+    "skill_used": "fabrication",
+    "difficulty": 2,
+    "autolearn": true,
+    "using": [ [ "forging_standard", 1 ], [ "bronzesmithing_tools", 1 ] ],
+    "tools": [ [ [ "casting_mold", -1 ] ] ],
+    "components": [ [ [ "scrap_aluminum", 3780 ] ] ]
```
Comment (Karol1223) [inline data/json/recipes/other/materials.json:358]: Why are you making zinc out of aluminum? We don't have scrap zinc for you to make this recipe with. Unless you want to go with the atomic weight item... which given the amount I'm assuming was the intention.

File: data/json/recipes/other/materials.json [hunk 2]
```diff
@@ -137,6 +177,45 @@
...
+    "type": "recipe",
+    "activity_level": "LIGHT_EXERCISE",
+    "category": "CC_OTHER",
+    "subcategory": "CSC_OTHER_MATERIALS",
+    "skill_used": "fabrication",
+    "difficulty": 2,
+    "autolearn": true,
+    "using": [ [ "forging_standard", 1 ], [ "bronzesmithing_tools", 1 ] ],
+    "tools": [ [ [ "casting_mold", -1 ] ] ],
+    "components": [ [ [ "scrap_copper", 34 ] ] ]
```
Comment (Karol1223) [inline data/json/recipes/other/materials.json:217]:
```suggestion
    "components": [ [ [ "scrap_copper", 36 ] ] ]
```
Double check masses when you do this stuff.

File: data/json/recipes/other/materials.json [hunk 3]
```diff
@@ -105,6 +105,46 @@
...
+    "activity_level": "LIGHT_EXERCISE",
+    "category": "CC_OTHER",
+    "subcategory": "CSC_OTHER_MATERIALS",
+    "skill_used": "fabrication",
+    "difficulty": 2,
+    "autolearn": true,
+    "using": [ [ "forging_standard", 1 ], [ "bronzesmithing_tools", 1 ] ],
+    "tools": [ [ [ "casting_mold", -1 ] ] ],
+    "//": "There is a slight loss here, but rounding up would mean you could create more magically.",
+    "components": [ [ [ "scrap_bronze", 25 ] ] ]
```
Comment (Karol1223) [inline data/json/recipes/other/materials.json:146]:
```suggestion
    "//": "There is a slight loss here, but rounding up would mean you could create more magically.",
    "components": [ [ [ "scrap_bronze", 26 ] ] ]
```
You were literally generating mass, not losing it.

File: data/json/recipes/other/materials.json [hunk 4]
```diff
@@ -940,6 +1058,46 @@
...
+    "activity_level": "LIGHT_EXERCISE",
+    "category": "CC_OTHER",
+    "subcategory": "CSC_OTHER_MATERIALS",
+    "skill_used": "fabrication",
+    "difficulty": 2,
+    "autolearn": true,
+    "using": [ [ "forging_standard", 1 ], [ "bronzesmithing_tools", 1 ] ],
+    "tools": [ [ [ "casting_mold", -1 ] ] ],
+    "//": "There is a slight loss here, but rounding up would mean you could create more magically.",
+    "components": [ [ [ "scrap_aluminum", 11 ] ] ]
```
Comment (Karol1223) [inline data/json/recipes/other/materials.json:1099]:
```suggestion
    "components": [ [ [ "scrap_aluminum", 54 ] ] ]
```
How were you calculating this? Did you assume chunks not scrap?

File: data/json/uncraft/generic.json [hunk 5]
```diff
@@ -311,10 +311,18 @@
...
+  },
+  {
+    "result": "1l_zinc",
+    "type": "uncraft",
+    "activity_level": "MODERATE_EXERCISE",
     "time": "1 h",
     "tools": [ [ [ "angle_grinder", 50 ] ] ],
-    "//": "perfect = 20000; subtracted 5.5% loss",
-    "components": [ [ [ "zinc_metal", 18900 ] ] ]
+    "//": "[was] perfect = 20000; subtracted 5.5% loss",
```
Comment (Karol1223) [inline data/json/uncraft/generic.json:324]: Just remove this comment at that point. This is only going to confuse people..

**Review by Karol1223 — CHANGES_REQUESTED** [review]

File: data/json/recipes/other/materials.json (hunk 1 above)
Comment (juur) [reply data/json/recipes/other/materials.json:358]: That was a copy/paste failure. It should read "zinc_metal", I hope that's OK?

Comment (Karol1223) [reply data/json/recipes/other/materials.json:358]: In this case you'd be looking at 3978. Not sure how you got this number.

**Comment by Procyonae** [comment]
Why is this useful?

**Comment by juur** [comment]
> Why is this useful?

I believe that the scrap variants are space inefficient vs. ingots. Having 50,000 scrap pieces of bronze vs. a much smaller number of ingots is surely more realistic and would take up far less volume?

File: data/json/items/ammo.json [hunk 6]
```diff
@@ -1695,6 +1695,24 @@
...
+    "volume": "38 ml",
+    "//2": "appears to be about 20% cheaper than bronze",
+    "price": 140,
+    "price_postapoc": 80,
+    "to_hit": -2,
+    "material": [ "brass" ],
+    "symbol": ",",
+    "color": "yellow",
+    "melee_damage": { "bash": 1 }
+  },
```
Comment (Karol1223) [inline data/json/items/ammo.json:1715]: Throw it in the metal resource file. We don't need more non-ammo in ammo files.

**Review by Karol1223 — CHANGES_REQUESTED** [review]

**Comment by Procyonae** [comment]
> I believe that the scrap variants are space inefficient vs. ingots. Having 50,000 scrap pieces of bronze vs. a much smaller number of ingots is surely more realistic and would take up far less volume?

Can you actually use them without reducing them down again though?
What does recipe time default to bc you don't seem to have defined it for any of the ingot forming recipes?
You should probably add side lengths too bc ingots aren't cubes

**Comment by juur** [comment]
> > I believe that the scrap variants are space inefficient vs. ingots. Having 50,000 scrap pieces of bronze vs. a much smaller number of ingots is surely more realistic and would take up far less volume?
> 
> Can you actually use them without reducing them down again though? What does recipe time default to bc you don't seem to have defined it for any of the ingot forming recipes? You should probably add side lengths too bc ingots aren't cubes

I didn't make the ingot items, they're already in without side lengths. I will add some durations to the recipe, good check.

File: data/json/recipes/other/materials.json [hunk 7]
```diff
@@ -105,6 +105,241 @@
...
+    "type": "recipe",
+    "activity_level": "LIGHT_EXERCISE",
+    "category": "CC_OTHER",
+    "subcategory": "CSC_OTHER_MATERIALS",
+    "skill_used": "fabrication",
+    "difficulty": 2,
+    "autolearn": true,
+    "using": [ [ "forging_standard", 1 ], [ "bronzesmithing_tools", 1 ] ],
+    "tools": [ [ [ "casting_mold", -1 ] ] ],
+    "components": [ [ [ "lead", 2268 ] ] ]
```
Comment (Karol1223) [inline data/json/recipes/other/materials.json:301]:
```suggestion
    "components": [ [ [ "lead", 2269 ] ] ]
```
You keep rounding down. Don't.

File: data/json/recipes/other/materials.json [hunk 8]
```diff
@@ -105,6 +105,241 @@
...
+    "type": "recipe",
+    "activity_level": "LIGHT_EXERCISE",
+    "category": "CC_OTHER",
+    "subcategory": "CSC_OTHER_MATERIALS",
+    "skill_used": "fabrication",
+    "difficulty": 2,
+    "autolearn": true,
+    "using": [ [ "forging_standard", 5 ], [ "bronzesmithing_tools", 1 ] ],
+    "tools": [ [ [ "casting_mold", -1 ] ] ],
+    "components": [ [ [ "scrap_brass", 29 ] ] ]
```
Comment (Karol1223) [inline data/json/recipes/other/materials.json:145]:
```suggestion
    "components": [ [ [ "scrap_brass", 30 ] ] ]
```

**Review by Karol1223 — CHANGES_REQUESTED** [review]

**Review by kasanryukin — CHANGES_REQUESTED** [review]
None of these should be autolearn, and should all be in a textbook, specifically brass and bronze should be in the alloying manual, and the ingot casting for the rest should be moved into a book.  Speaking from years of personal experience, casting metal isn't as simple heating until liquid and pouring in a mold. There are several steps that get hand waved in game, but are important if you want an ingot that doesn't contain a large amount of impurities.

Additionally, the melting point of platinum is nearly 1,000 degrees F / 537 C higher than iron and requires specialized equipment that is very, very uncommon.  (Edit: And you should not use acetylene ever with platinum. It's too much of a carborizing flame)

**Comment by juur** [comment]
> None of these should be autolearn, and should all be in a textbook, specifically brass and bronze should be in the alloying manual, and the ingot casting for the rest should be moved into a book. Speaking from years of personal experience, casting metal isn't as simple heating until liquid and pouring in a mold. There are several steps that get hand waved in game, but are important if you want an ingot that doesn't contain a large amount of impurities.
> 
> Additionally, the melting point of platinum is nearly 1,000 degrees F / 537 C higher than iron and requires specialized equipment that is very, very uncommon. (Edit: And you should not use acetylene ever with platinum. It's too much of a carborizing flame)

Thank you for the feedback. I think I've implemented some of it.
I have been using "forging_standard" - would you say this is reasonable (taking your hand waving point into consideration) for most of the metals/alloys being melted - with the exception of platinum?

Is there anything in game currently that could be used for platinum, with some level of hand waving? :-)

**Comment by kasanryukin** [comment]
> > None of these should be autolearn, and should all be in a textbook, specifically brass and bronze should be in the alloying manual, and the ingot casting for the rest should be moved into a book. Speaking from years of personal experience, casting metal isn't as simple heating until liquid and pouring in a mold. There are several steps that get hand waved in game, but are important if you want an ingot that doesn't contain a large amount of impurities.
> > Additionally, the melting point of platinum is nearly 1,000 degrees F / 537 C higher than iron and requires specialized equipment that is very, very uncommon. (Edit: And you should not use acetylene ever with platinum. It's too much of a carborizing flame)
> 
> Thank you for the feedback. I think I've implemented some of it. I have been using "forging_standard" - would you say this is reasonable (taking your hand waving point into consideration) for most of the metals/alloys being melted - with the exception of platinum?
> 
> Is there anything in game currently that could be used for platinum, with some level of hand waving? :-)

There are high temperature arc furnaces in the game that would do it easily.  You'd need a specialized ceramic crucible (which doesn't exist in game) since a graphite, iron, or clay crucible would melt or spall before the platinum melted.

**Comment by juur** [comment]
> There are high temperature arc furnaces in the game that would do it easily. You'd need a specialized ceramic crucible (which doesn't exist in game) since a graphite, iron, or clay crucible would melt or spall before the platinum melted.

Shall I remove the recipes for platinum ingots for now, so the rest can go through? I did a bit of googling and found various examples of quartz and graphite moulds/crucibles for platinum, but how you might craft these is very complicated, so unsure if just having them rare finds is/isn't the right thing to do.

**Comment by juur** [comment]
> > There are high temperature arc furnaces in the game that would do it easily. You'd need a specialized ceramic crucible (which doesn't exist in game) since a graphite, iron, or clay crucible would melt or spall before the platinum melted.
> 
> Shall I remove the recipes for platinum ingots for now, so the rest can go through? I did a bit of googling and found various examples of quartz and graphite moulds/crucibles for platinum, but how you might craft these is very complicated, so unsure if just having them rare finds is/isn't the right thing to do.

I've removed the recipes for platinum ingots now, so hopefully this can be merged :-)

**Comment by Procyonae** [comment]
You don't need to bump the tests when all the required tests have passed

**Comment by Karol1223** [comment]
Well, *now* you do. Because the required tests failed 🤣 

But if you have the ``BasicBuildPassed`` label don't bump the tests.

**Comment by juur** [comment]
> Well, _now_ you do. Because the required tests failed 🤣
> 
> But if you have the `BasicBuildPassed` label don't bump the tests.

I can't win :-)
Useful to know what tags matter though, thank you.

**Comment by juur** [comment]
Is this OK to merge now?

**Review by Karol1223 — APPROVED** [review]
It *looks* good to merge for me. Unless I'm missing something.

**Review by kasanryukin — APPROVED** [review]
Everything looks great to me. I'd say ready to merge.  Ping it in the ready-to-merge channel on discord.

**Comment by juur** [comment]
> Everything looks great to me. I'd say ready to merge. Ping it in the ready-to-merge channel on discord.

Would love to, but not used Discord for several years and it says I don't have permission to :-)

**Comment by kasanryukin** [comment]
> > Everything looks great to me. I'd say ready to merge. Ping it in the ready-to-merge channel on discord.
> 
> Would love to, but not used Discord for several years and it says I don't have permission to :-)

I've posted it for you then.  Be sure to poke somebody in the #meta channel to get the contributor role so you can post there when your PRs are good to go.

**Pull request merged**